### ISSUE TITLE
Scapegoat results not reported in Sonar

### ISSUE THREAD

**Issue by h-ngo** [mwz/sonar-scala#211]
@mwz
This issue was previously reported and fixed https://github.com/mwz/sonar-scala/pull/96

But currently encountered this issue in multi-module Gradle project. According to the logs (see below with masked module/class named due to business restrictions), it looks like the reported was loaded, but issues are not reported as code smell in Sonar (there are violations in the html that scapegoat generated).

Any ideas? Thanks in advance :)

sonarqube: 6.7.7 - sonar-scala: 6.8.0 - Scapegoat profile was selected as default

Logs (file not found errors are Spec files. No other scapegoat related errrors):

09:27:44.626 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Loading the scapegoat report file: '/export/home/dev/IdeaProjects/abc-app-scapegoat/abc-business/build/scapegoat/scapegoat.xml'.
09:27:44.626 [DEBUG] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The current working directory is: '/export/home/dev/IdeaProjects/abc-app-scapegoat'.
09:27:44.631 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Successfully loaded the scapegoat report file.
09:27:44.631 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/MS/MSFlowSpec.scala'.
09:27:44.632 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/MS/MSFlowSpec.scala' couldn't be found.
09:27:44.632 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/MP/MPFlowSpec.scala'.
09:27:44.632 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/MP/MPFlowSpec.scala' couldn't be found.
09:27:44.632 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/DS/DSFlowSpec.scala'.
09:27:44.633 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/DS/DSFlowSpec.scala' couldn't be found.
09:27:44.633 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/MP/MPValidatorSpec.scala'.
09:27:44.633 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/MP/MPValidatorSpec.scala' couldn't be found.
09:27:44.633 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/MU/MUValidatorSpec.scala'.
09:27:44.634 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/MU/MUValidatorSpec.scala' couldn't be found.
09:27:44.634 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/DR/DRFlowSpec.scala'.
09:27:44.634 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/DR/DRFlowSpec.scala' couldn't be found.
09:27:44.634 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/MU/MUFlowSpec.scala'.
09:27:44.635 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/MU/MUFlowSpec.scala' couldn't be found.
09:27:44.635 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/abc/flow/DS/DSRollbackSpec.scala'.
09:27:44.635 [ERROR] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] The file 'src/test/scala/abc/flow/DS/DSRollbackSpec.scala' couldn't be found.
09:27:44.635 [INFO] [org.sonarqube.gradle.SonarQubeTask] Sensor Scapegoat Sensor [scala] (done) | time=9ms

**Comment by mwz** [comment]
Hi @h-ngo, we currently don't support running analysis on test files, neither does scapegoat (see https://github.com/sksamuel/sbt-scapegoat/issues/46).

Feel free to upvote #141 and #142 if that's something you would like to see implemented in the future.

**Comment by h-ngo** [comment]
@mwz 
Sorry for the confusion. Meant to say to ignore error on analysis on test files.

For analysis on non-test files, we saw violations in the html that scapegoat generated

![image](https://user-images.githubusercontent.com/34242287/58732567-e413b600-83bf-11e9-81e3-c9107d812898.png)

and was able to see sonarqube loaded the reported successfully.

09:27:44.631 [INFO] [org.sonarqube.gradle.SonarQubeTask] [sonar-scala-scapegoat] Successfully loaded the scapegoat report file.

But these violations are not reported in Sonar as code smells.

**Comment by mwz** [comment]
Is there anything in the logs about those files? All of the errors from your log above are related to the test files.

**Comment by h-ngo** [comment]
There's no log about non-test files. In fact, I just re-examined the html report. It actually reported on test files only. So, I guess our Gradle configuration is incorrect.

Closing this ticket for now.... Thanks a lot for your quick responses, @mwz !!

**Comment by mwz** [comment]
👍 

I agree it doesn't seem right that Scapegoat is reporting issues about test files.
Have a look at our example multi-module Gradle project [here](https://github.com/mwz/sonar-scala/tree/master/examples/gradle/multi-module) for comparison.

**Comment by cosminci** [comment]
@mwz We're seeing the same issue, but no scapegoat results seem to be reported at all in SonarQube (that is, nothing for source files either, not just test files). 

Here are the logs:
```
[INFO] [sonar-scala-scapegoat] Initializing the Scapegoat sensor.
[INFO] [sonar-scala-scapegoat] Loading the scapegoat report file: '<path_to_parent_pom>/<maven_submodule>/target/scapegoat.xml'.
[INFO] [sonar-scala-scapegoat] Successfully loaded the scapegoat report file.
[INFO] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/<path>...Test.scala'.
[ERROR] [sonar-scala-scapegoat] The file 'src/test/scala/<path>Test.scala' couldn't be found.
```
Looking into the XML report there are quite a few issues for source files reported by scapegoat, but none of them end up in SonarQube. The scala style issues do appear.

Any ideas what might be wrong?

**Comment by mwz** [comment]
Hi @cosminci, you're not executing a scan on Windows, are you?

>The file 'src/test/scala/\<path>Test.scala' couldn't be found.

Just to clarify, the `<path>` part in the above file path isn't actually part of the path, I'm assuming you just shortened/obfuscated it before you posted it?

It might be worth trying to run it with `-Dsonar.verbose=true` to see the debug info - there might be something that can give us some clues.

**Comment by cosminci** [comment]
Running on both mac & docker alpine and results are the same. The path is obfuscated yes, it's just the usual package path e.g. `src/test/scala/com/cosminci/testproject/RandomTest.scala`.

Weirdly enough, I'm not seeing any DEBUG or TRACE logs when starting with `-Dsonar.verbose=true` or `-Dsonar.log.level=TRACE` or setting either of these options as properties inside the pom with the rest of the sonar properties e.g.:

```
<properties>
    <sonar.host.url>http://localhost</sonar.host.url>
    <sonar.log.level>TRACE</sonar.log.level>
    <sonar.scala.version>${scala.version}</sonar.scala.version>
    <sonar.core.codeCoveragePlugin>scoverage</sonar.core.codeCoveragePlugin>
    <sonar.junit.reportPaths>${project.build.directory}/surefire-reports</sonar.junit.reportPaths>    <sonar.scala.scoverage.reportPath>${project.build.directory}/scoverage.xml</sonar.scala.scoverage.reportPath>
    <sonar.scala.scapegoat.reportPath>${project.build.directory}/scapegoat.xml</sonar.scala.scapegoat.reportPath>
</properties>
```

`sonar-maven-plugin` version is latest - 3.7.0.1746

`SonarQube` is run as a local docker container via the image you provide: `mwizner/sonarqube-scala-plugins latest-full` (`[INFO] SonarQube version: 8.1.0`), but I get the same results with an already provisioned LTS version which has the corresponding LTS sonar-scala plugin installed manually.

Leaving the `*Test` file error messages aside, the logs seem to point out everything is ok:
```
[INFO] Sensor Scapegoat Sensor [scala]
[INFO] [sonar-scala-scapegoat] Initializing the Scapegoat sensor.
[INFO] [sonar-scala-scapegoat] Loading the scapegoat report file: '<obfuscated>/target/scapegoat.xml'.
[INFO] [sonar-scala-scapegoat] Successfully loaded the scapegoat report file.
[INFO] Sensor Scapegoat Sensor [scala] (done) | time=13ms
``` 
This happens for every module in the multi-module project. Similarly sensors for scala style & scoverage are run, and their results are available in the UI. It sounds like a load / display issue in SonarQube. 

I haven't tinkered with any settings in the server e.g. default quality profile for scala is still `Recommended by sonar-scala`

**Comment by mwz** [comment]
>[INFO] [sonar-scala-scapegoat] Saving the scapegoat issues for file 'src/test/scala/<path>...Test.scala'.
[ERROR] [sonar-scala-scapegoat] The file 'src/test/scala/<path>Test.scala' couldn't be found.

This path doesn't look right to me:
1. Scapegoat shouldn't run against your test files, just source files, but it looks like somehow you ended up with Scapegoat results for your test files, i.e. from `src/test/scala`?
2. When you look at the [logs](https://app.circleci.com/pipelines/github/mwz/sonar-scala/652/workflows/497b54bf-2987-475c-8090-a653ba8c475d/jobs/1526/parallel-runs/0/steps/0-105) from running sonar-scanner against a multi-module maven project, you'll notice that there is no source directory prefix in log messages when saving scapegoat issues, so the above log messages also don't look correct to me. You should expect to see the the path relative to the sources folder, e.g. `com/cosminci/File.scala`. Can I double-check that you have your source files inside `src/main/scala` within each module?

This looks like a misconfiguration to me - if you could provide a minimal example project example with this issue I'm more than happy to have a look at it and debug sonar-scala if it turns out the issue is on our end.

**Comment by cosminci** [comment]
Decided to start with the [example project](https://github.com/mwz/sonar-scala/tree/master/examples/mvn/multi-module) and then gradually evolve it towards the more complex pom structure I have to see where they start diverging.

The first thing I noticed is that scapegoat code smells don't end up in SonarQube for the example either, so something likely changed since the example was last updated. 

Regarding the two points you made, they're probably related - I only see logs for test files and I don't see them when building the example project. I should be able to find out the reason by doing what I mentioned above.

**Comment by mwz** [comment]
> The first thing I noticed is that scapegoat code smells don't end up in SonarQube for the example either, so something likely changed since the example was last updated.

How do you mean? [Here](https://sonar.sonar-scala.com/project/issues?id=sonar-scala%3Aexample-mvn-multi-module&issues=AWllABOFfsoN-mdhPCwN&open=AWllABOFfsoN-mdhPCwN) you can see an example scapegoat issue reported for the mvn multi-module project. It would be good if you could share what you're doing different to how the example is set up. Perhaps you're using new versions of dependencies and something causes it to break?

**Comment by cosminci** [comment]
Rechecked and I do see scapegoat issues included in the example project, must've been confused by the Scalastyle tag of the issues I checked initially.

I've updated the example pom step by step to match my own and it looks like the problem is due to the scala version. Steps:
1. Set
```
        <scala.version>2.13.2</scala.version>
        <scala.binary.version>2.13</scala.binary.version>
```
This fails compilation: `[ERROR] error: java.lang.NoSuchMethodError: scala.tools.nsc.Global.reporter()Lscala/tools/nsc/reporters/Reporter;` 

Tried running it under JDK8 & JDK13 with the same result.

2. Update the scapegoat plugin to 1.4.4, and change its dependencies to use the full scala version (starting with 1.4.x, scapegoat is released for each minor version). The result is that the example project now only has 9 issues reported in Sonar, all of which are scalastyle.

Regarding the fact my scapegoat config is also running for test files - I'm really stumped on this one as the project structures and pom configs are almost identical. It looks like after running on the test files, it overrides the proper source file report with test issues, which don't end up in SQ. I'll restart this track and create some gists once we get to the bottom of the scala 2.13 support.

**Comment by cosminci** [comment]
Removed the sonar-scala docker image I had locally and downloaded the 8.2 one, and the scapegoat reports are there \o/. 

Also, in order to fix the overriding reports it was necessary not just to create separate executions for the `compile` and `testCompile` phases, but also to move the scapegoat plugin's configuration to only be inside the execution for the `compile` phase. It's still a mystery to me how scapegoat is not running against the test files in the example project.

Thanks for taking the time to provide valuable hints and for the great work with the plugin!
